packages/transfer/src/index.vue

```diff
@@ -0,0 +1,338 @@
+    )
+
+    const panelFilterPlaceholder = computed(
+      () => props.filterPlaceholder || t('el.transfer.filterPlaceholder'),
+    )
+
+    watch(
+      () => props.modelValue,
+      val => {
+        emit(UPDATE_MODEL_EVENT, val)
+      },
+    )
+
+    provide('defaultScopedSlots', computed(() => slots.default))
+
+    const {
+      leftChecked,
+      rightChecked,
+    } = toRefs(initData)
+
+    return {
+      sourceData,
+      targetData,
+      onSourceCheckedChange,
+      onTargetCheckedChange,
+      addToLeft,
+      addToRight,
+
+      leftChecked,
+      rightChecked,
+
+      hasButtonTexts,
+      leftPanelTitle,
+      rightPanelTitle,
+      panelFilterPlaceholder,
+      clearQuery,
+    }
+  },
+})
+
+const useData = props => {
+  const propsKey = computed(() => {
+    return props.props.key
+  })
+
+  const dataObj = computed(() => {
+    return props.data.reduce(
+      (o, cur) => (o[cur[propsKey.value]] = cur) && o,
+      {},
+    )
+  })
+
+  const sourceData = computed(() => {
+    return props.data.filter(
+      item => props.modelValue.indexOf(item[propsKey.value]) === -1,
+    )
+  })
+
+  const targetData = computed(() => {
+    if (props.targetOrder === 'original') {
+      return props.data.filter(
+        item => props.modelValue.indexOf(item[propsKey.value]) > -1,
+      )
+    } else {
+      return props.modelValue.reduce((arr, cur) => {
+        const val = dataObj.value[cur]
+        if (val) {
+          arr.push(val)
+        }
+        return arr
+      }, [])
+    }
+  })
+
+  return {
+    propsKey,
+    sourceData,
+    targetData,
+  }
+}
+
+const useCheckedChange = (initData, emit) => {
+  const onSourceCheckedChange = (val, movedKeys) => {
+    initData.leftChecked = val
+    if (movedKeys === undefined) return
+    emit(LEFT_CHECK_CHANGE_EVENT, val, movedKeys)
+  }
+
+  const onTargetCheckedChange = (val, movedKeys) => {
+    initData.rightChecked = val
+    if (movedKeys === undefined) return
+    emit(RIGHT_CHECK_CHANGE_EVENT, val, movedKeys)
+  }
+
+  return {
+    onSourceCheckedChange,
+    onTargetCheckedChange,
+  }
+}
+
+const useMove = (props, initData, propsKey, emit) => {
+  const _emit = (value, type: 'left' | 'right', checked: string[]) => {
+    emit(UPDATE_MODEL_EVENT, value)
+    emit(CHANGE_EVENT, value, type, checked)
+  }
+
+  const addToLeft = () => {
+    let currentValue = props.modelValue.slice()
+    initData.rightChecked.forEach(item => {
+      const index = currentValue.indexOf(item)
+      if (index > -1) {
+        currentValue.splice(index, 1)
+      }
+    })
+    _emit(currentValue, 'left', initData.rightChecked)
+  }
+
+  const addToRight = () => {
+    let currentValue = props.modelValue.slice()
+    const itemsToBeMoved = []
+    props.data.forEach(item => {
+      const itemKey = item[propsKey.value]
+      if (
+        initData.leftChecked.indexOf(itemKey) > -1 &&
+        props.modelValue.indexOf(itemKey) === -1
```

Comment:
`initData.leftChecked` and `props.modelValue`
They seem to be Array, why not use includes ?

Reply:
Yes, I'll use `filter` to re-write this loop in next commit.